Sources/Core/AWSSDKIdentity/Sources/AWSSDKIdentity/AWSCredentialIdentityResolvers/IMDSAWSCredentialIdentityResolver.swift

```diff
@@ -6,19 +6,202 @@
+            configValue: nil,
+            envVarName: "AWS_EC2_METADATA_DISABLED",
+            configFieldName: "disable_ec2_metadata",
+            fileBasedConfig: fileBasedConfig,
+            converter: { Bool($0.lowercased()) }
+        ).value ?? false
+        guard !imdsDisabled else {
+            throw AWSCredentialIdentityResolverError.failedToResolveAWSCredentials(
+                "IMDSAWSCredentialIdentityResolver: "
+                + "IMDS is disabled via env var or shared config file."
             )
+        }
+    }
+
+    private func resolveCredentialURLBase() -> String {
+        return apiVersion == .legacy
+        ? "/latest/meta-data/iam/security-credentials/"
+        : "/latest/meta-data/iam/security-credentials-extended/"
+    }
+
+    private func parseJSONCredentials(
+        jsonCredentialString: String
+    ) throws -> AWSCredentialIdentity {
+        guard let jsonCredentialData = jsonCredentialString.data(using: .utf8) else {
+            throw IMDSError.deserializationError(
+                "IMDSAWSCredentialIdentityResolver: "
+                + "Failed to decode JSON credential response from IMDS."
+            )
+        }
+        let jsonCredentials = try JSONDecoder().decode(
+            JSONCredentialResponse.self,
+            from: jsonCredentialData
+        )
+        return AWSCredentialIdentity(
+            accessKey: jsonCredentials.accessKeyID,
+            secret: jsonCredentials.secretAccessKey,
+            accountID: jsonCredentials.accountID,
+            expiration: jsonCredentials.expiration,
+            sessionToken: jsonCredentials.token
         )
     }
 }
+
+private struct JSONCredentialResponse: Codable {
+    let code: String
+    let lastUpdated: Date
+    let type: String
+    let accessKeyID: String
+    let secretAccessKey: String
+    let token: String
+    let expiration: Date
+    let accountID: String?
+
+    enum CodingKeys: String, CodingKey {
+        case code = "Code"
+        case lastUpdated = "LastUpdated"
+        case type = "Type"
+        case accessKeyID = "AccessKeyId"
+        case secretAccessKey = "SecretAccessKey"
+        case token = "Token"
+        case expiration = "Expiration"
+        case accountID = "AccountId"
+    }
+
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        let formatter = ISO8601DateFormatter()
+
+        code = try container.decode(String.self, forKey: .code)
+        lastUpdated = formatter.date(from: try container.decode(String.self, forKey: .lastUpdated))!
+        type = try container.decode(String.self, forKey: .type)
+        accessKeyID = try container.decode(String.self, forKey: .accessKeyID)
+        secretAccessKey = try container.decode(String.self, forKey: .secretAccessKey)
+        token = try container.decode(String.self, forKey: .token)
+        expiration = formatter.date(from: try container.decode(String.self, forKey: .expiration))!
```

Comment:
Let's throw an error instead of force-unwrap here (and a few lines above as well.)
`ISO8601DateFormatter` has some flexibility around formats, but it's not unheard of for devs to be sloppy with date formats & I don't want to crash on this.

Reply:
Changed as suggested.